cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/PreviewDataSubscriberService.java

```diff
@@ -173,6 +177,20 @@ protected PreviewMessage decodeMessage(Message message) {
     return GSON.fromJson(message.getPayloadAsString(), PreviewMessage.class);
   }
 
+  /**
+   * Emit the preview run time metric.
+   */
+  private void emitRunTimeMetric(PreviewStatus previewStatus, ApplicationId applicationId) {
+    long runTime = (previewStatus.getEndTime() - previewStatus.getStartTime()) / 1000;
+    Map<String, String> tags = ImmutableMap.<String, String>builder()
+      .put(Constants.Metrics.Tag.NAMESPACE, NamespaceId.SYSTEM.getNamespace())
+      .put(Constants.Metrics.Tag.APP, applicationId.getApplication())
```

Comment:
Metrics context that involves execution should have program and run id. See `AbstractContext` for reference

https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/AbstractContext.java#L293

Reply:
done

Reply:
You need program type as well since program name can be the same for different program type. Please do it exactly like the one in AbstractContext. It maybe good to refactor that method out to a common util.

Reply:
fixed.